Expand parameters to support the Array type.

EliZucker opened the pull request:
<!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

# Changes

### **Introduction:**

This change was a lot harder than I initially thought it would be! There were a lot of unexpected design considerations and edge cases to catch. 

I'm not sure how to logically divide this PR, since all of the test refactoring depends on the new ArrayOrString type so can't be in a separate commit. 

Please let me know if you have any suggestions, since this is my first PR on Tekton of this size!

~~This PR is technically incomplete because I want to make the user documentation as clear as possible. However, the actual code for funtionality/tests is in a separate commit and ready for review.~~ **PR is completely ready for review**. Thanks!

### **Commit message:**

Expand parameters to support the Array type.

Instead of just strings, add support for users to supply parameters in the form of an array. Also include appropriate type validation and defaulting to support backwards-compatibility. Fixes #207.

The size of a user-supplied array is not specified anywhere in the corresponding `ParamSpec`, which is useful for defining tasks that require a dynamic number of strings used as distinct elements in an array. For instance, one might use an array parameter in an image-building task that feeds in a dynamic number of flags via the `args` field for a particular step.

The implementation is defined such that an Array parameter can only be used if the replacement string is completely isolated and within a field that is an array of strings (currently eligible fields are `command` and `args`). For instance, the webhook will prevent an array parameter from being used in the `image` field. Similarly, an array parameter used in a composite string, like `"value=${array}"`, is invalid and will be caught. The decision was made to completely prevent any use of array parameters as strings because it clutters Tekton with unnecessary functionality and is naturally extensible beyond the intended scope. For instance, if the behavior was defined such that array parameters used in the context of strings were automatically converted by separating them with a space, then that relatively arbitrary implementation could be questioned and eventually lead to needing to specify a custom delimiter (like commas), and so on.

Another implementation detail to note is that the `ArrayOrString` type was necessary, in combination with implementing json.Marshaller, to support backwards-compatibility with strings. This is based off `IntOrString` from kubernetes/apimachinery, and allows the value-type to be immediately parsed and decided through the webhook.

Lastly, a possibly unintuitive design decision is that if an EMPTY array parameter properly replaces an array element, then that string in the original array will be completely removed.

### **Example YAML (not in commit message):**

Feel free to deploy pipelines with this PR's code changes and try out array parameters for yourself! Here is an example YAML that should run:

```
apiVersion: tekton.dev/v1alpha1
kind: Task
metadata:
  name: build-push
  namespace: default
spec:
  inputs:
    params:
      - name: pathToContext
        type: array
  steps:
    - name: printparam
      image: ubuntu
      command: [echo]
      args: ["${inputs.params.pathToContext}"]
---
apiVersion: tekton.dev/v1alpha1
kind: Pipeline
metadata:
  name: pipeline-with-array
spec:
  params:
    - name: context
      description: Path to context
      default: 
        - "this"
        - "should"
        - "be overridden"
  tasks:
    - name: build-skaffold-web
      taskRef:
        name: build-push
      params:
        - name: pathToContext
          value: 
            - "1"
            - "${params.context}"
            - "2"
---
apiVersion: tekton.dev/v1alpha1
kind: PipelineRun
metadata:
  name: pipelinerun-with-array
spec:
  pipelineRef:
    name: pipeline-with-array
  params:
    - name: context
      value: 
        - "element1"
        - "element2"
```

After running this, the step's pod reads the following args:
```
echo
      --
      1
      element1
      element2
      2
```
# Submitter Checklist

These are the criteria that every PR should meet, please check them off as you
review them:

- [x] Includes [tests](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if functionality changed/added)
- [x] Includes [docs](https://github.com/tektoncd/community/blob/master/standards.md#principles) (if user facing)
- [x] Commit messages follow [commit message best practices](https://github.com/tektoncd/community/blob/master/standards.md#commit-messages)

_See [the contribution guide](https://github.com/tektoncd/pipeline/blob/master/CONTRIBUTING.md) for more details._

Double check this list of stuff that's easy to miss:

- If you are adding [a new binary/image to the `cmd` dir](../cmd), please update
  [the release Task](../tekton/publish.yaml) and [TaskRun](../tekton/publish-run.yaml) to build and release this image

# Release Notes

```
Parameters now support the Array type.
```

EliZucker commented:
/hold

EliZucker commented:
/test pull-tekton-pipeline-build-tests

EliZucker commented:
/retest

EliZucker commented:
/retest

EliZucker commented:
/retest

sbwsg commented:
This may need corollary work in the CLI tool, [tkn](https://github.com/tektoncd/cli), or perhaps just a new version to be cut with this updated code? I just installed your controller and applied your example YAML, then ran `tkn pipelineruns list`.  Received the following error:

```bash
~/dev/go/src/github.com/tektoncd/manual-tests/array-or-string                                                                                   [12/118]
 λ tkn pipelinerun list                                                                                                                                             
Failed to list pipelineruns from default namespace                                                                                                                  
Error: v1alpha1.PipelineRunList.Items: []v1alpha1.PipelineRun: v1alpha1.PipelineRun.Spec: v1alpha1.PipelineRunSpec.Params: []v1alpha1.Param: v1alpha1.Param.Value: R
eadString: expects " or n, but found [, error found in #10 byte of ...|,"value":["element1"|..., bigger context ...|da"},"spec":{"params":[{"name":"context","value$
:["element1","element2"]}],"pipelineRef":{"name":"pi|...                                                                                                            
Usage:                                                                                                                                                              
  tkn pipelinerun list [flags]
```

So I think whatever tkn is using to unmarshall the JSON response from the kube api needs to be updated to take your array support into account.

vdemeester commented:
> This may need corollary work in the CLI tool, [tkn](https://github.com/tektoncd/cli), or perhaps just a new version to be cut with this updated code? I just installed your controller and applied your example YAML, then ran `tkn pipelineruns list`. Received the following error:
> 
> ```shell
> ~/dev/go/src/github.com/tektoncd/manual-tests/array-or-string                                                                                   [12/118]
>  λ tkn pipelinerun list                                                                                                                                             
> Failed to list pipelineruns from default namespace                                                                                                                  
> Error: v1alpha1.PipelineRunList.Items: []v1alpha1.PipelineRun: v1alpha1.PipelineRun.Spec: v1alpha1.PipelineRunSpec.Params: []v1alpha1.Param: v1alpha1.Param.Value: R
> eadString: expects " or n, but found [, error found in #10 byte of ...|,"value":["element1"|..., bigger context ...|da"},"spec":{"params":[{"name":"context","value$
> :["element1","element2"]}],"pipelineRef":{"name":"pi|...                                                                                                            
> Usage:                                                                                                                                                              
>   tkn pipelinerun list [flags]
> ```
> 
> So I think whatever tkn is using to unmarshall the JSON response from the kube api needs to be updated to take your array support into account.

Indeed, I'll create an issue in the cli :angel:

EliZucker commented:
/assign @imjasonh

afrittoli commented on pkg/apis/pipeline/v1alpha1/task_validation.go:
```diff
@@ -173,6 +213,47 @@ func validateResourceVariables(steps []corev1.Container, inputs *Inputs, outputs
 	return validateVariables(steps, "resources", resourceNames)
 }
 
+func validateArrayUsage(steps []corev1.Container, prefix string, vars map[string]struct{}) *apis.FieldError {
```
Woah, I didn't think about this :) Is this a complete list of all the step fields that we need to check?
Hopefully if we ever change the spec of Steps, we'll remember to update the validation here too.

EliZucker replied:
I replied to this 3 days ago and then eventually resolved it, but apparently my reply was either wiped or never sent :angry: 

I believe that is a complete list of all step fields that need to be checked; the step field checking logic is directly copied from `validateVariables()`. 

Something to consider is that in the future, if the spec is changed, someone might remember to update `validateVariables()` but forget to add that extra field to `validateArrayUsage()`. I don't believe those 2 methods can be combined, though, because `validateArrayUsage()` does distinctly different checks on each field in comparison to `validateVariables()`, like `validateTaskNoArrayReferenced` and `validateTaskArraysIsolated`. Additionally, `validateVariables()` is called in different contexts that wouldn't make sense to use array validation logic for (like resource validation).

EliZucker replied:
I just realized that each method is checking the fields of `corev1.container`, not a type we define ourselves, so I don't think it's as necessary to add an reminder comment. Let me know if you disagree.

ImJasonH commented on pkg/apis/pipeline/v1alpha1/param_types_test.go:
```diff
@@ -0,0 +1,147 @@
+		name            string
+		before          *v1alpha1.ParamSpec
+		defaultsApplied *v1alpha1.ParamSpec
+	}{
```
Let's collapse this whitespace

EliZucker replied:
done!

ImJasonH commented on pkg/apis/pipeline/v1alpha1/pipeline_validation_test.go:
```diff
@@ -118,16 +118,25 @@ func TestPipelineSpec_Validate(t *testing.T) {
 		p: tb.Pipeline("pipeline", "namespace", tb.PipelineSpec(
-			tb.PipelineParam("baz"),
-			tb.PipelineParam("foo-is-baz"),
+			tb.PipelineParam("baz", v1alpha1.ParamTypeString),
```
Similar to Andrea's comment above, we could make this test builder also assume the string type unless a type is passed as an option.

EliZucker replied:
That would definitely be better, but do you have any suggestions for how to do that in Go? Andrea's comment I think referenced the fact that the webhook will automatically set the ParamSpec's type field to "string"  if it is nil. 

I could make `PipelineParam(name string, types ...type)` a variadic function so that the type is optional, but then users of the builder can enter as many types as they want and the behavior is unclear (unless I manually check the # of additional parameters and return an error).

ImJasonH replied:
In that case you could either document that the last one wins, or, yeah, detect the presence of a previous type variadic arg and fail with a descriptive message if you see that. Either way is probably fine, so long as it's documented.

EliZucker replied:
So just checked the code : `func PipelineParam(name string, pt v1alpha1.ParamType, ops ...PipelineParamOp)` . Since the variadic part needs to be the final parameter (which is currently a list of operations), I'm not sure how to make ParamType optional.

EliZucker replied:
Could I guess make separate builder functions for each type of param..

ImJasonH commented on test/builder/pipeline_test.go:
```diff
@@ -121,10 +131,14 @@ func TestPipelineRun(t *testing.T) {
-				Name:  "first-param",
-				Value: "first-value",
-			}},
+			Params: []v1alpha1.ArrayOrStringParam{
```
collapse whitespace

EliZucker replied:
done!

ImJasonH commented on test/builder/pipeline_test.go:
```diff
@@ -121,10 +131,14 @@ func TestPipelineRun(t *testing.T) {
-				Name:  "first-param",
-				Value: "first-value",
-			}},
+			Params: []v1alpha1.ArrayOrStringParam{
```
Are we going to have to change the type of this `Params` field if we decide in the future we want to support `ArrayOrStringOrIntOrBoolOrStructParam`? :)

If so, maybe this indicates we should come up with a type name that doesn't tie us to the types available within it. I don't suppose we could reuse `Param`...?

EliZucker replied:
Retyping this comment because it looks like it never sent when I typed it 2 days ago:

Currently `Param` is an existing type used exclusively by `PipelineResources` and only can hold a string (used for things like a git url). I definitely like the idea of refactoring `ArrayOrStringParam`, though, because it is a more general parameter type than the current `Param` type is. 

I pushed a code change earlier that renames the existing `Param` type to `ResourceParam`, which I think is more specific and appropriate, and and renames `ArrayOrStringParam` to `Param`. 

Let me know if that change is good; one thing to note is that for future changes, the `Param` type name will never change, but its `Value` field, which is currently of type `ArrayOrString` would have to change to `ArrayOrStringOrIntOrBoolOrStruct`. I don't think that's avoidable.

ImJasonH replied:
Blech. I think we should try to figure out if there's some other (backward-compatible!) way to handle this should we decide we need to in future changes, but I don't want to block this PR on solving that.

The rest of that change lgtm.

EliZucker commented on docs/tasks.md:
```diff
@@ -537,8 +554,10 @@ spec:
   inputs:
     params:
     - name: package
+      type: string
```
Yeah you're right, the type is assumed to be "string" if nothing is specified, partly for backwards compatibility. I brought this up briefly at last week's meeting; I didn't know if we should encourage explicit parameter type definition for strings, and I think @vdemeester and some others said that they were in favor of encouraging explicit typing, even for strings. So I modified the readme examples to reflect that mindset.

afrittoli replied:
Thanks for the extra context, fine with me :)

EliZucker commented on pkg/apis/pipeline/v1alpha1/param_types.go:
```diff
@@ -30,11 +42,97 @@ type ParamSpec struct {
+// ArrayOrString is a type that can hold a single string or string array.
+// Used in JSON unmarshalling so that a single JSON field can accept
+// either an individual string or an array of strings.
+type ArrayOrString struct {
```
Thanks! It was based on [IntOrString](https://github.com/kubernetes/apimachinery/blob/master/pkg/util/intstr/intstr.go) which similarly uses json.unmarshall to allow for multiple formats in the YAML.

EliZucker commented:
/test pull-tekton-pipeline-integration-tests

sbwsg commented on docs/tasks.md:
````diff
@@ -394,6 +405,8 @@ To access an input parameter, replace `resources` with `params`.
 ${inputs.params.<name>}
 ```
 
+Note that for parameters of type `array`, each element of the array will be appended together separated by spaces when used in templating.
````
Is this still true?  Bit confused between this and the paragraph in your PR description:

> The implementation is defined such that an Array parameter can only be used if the replacement string is completely isolated and within a field that is an array of strings (currently eligible fields are command and args). For instance, the webhook will prevent an array parameter from being used in the image field. Similarly, an array parameter used in a composite string, like "value=${array}", is invalid and will be caught. The decision was made to completely prevent any use of array parameters as strings because it clutters Tekton with unnecessary functionality and is naturally extensible beyond the intended scope. For instance, if the behavior was defined such that array parameters used in the context of strings were automatically converted by separating them with a space, then that relatively arbitrary implementation could be questioned and eventually lead to needing to specify a custom delimiter (like commas), and so on.

I think I'm misunderstanding - arrays might still be concatenated together with spaces?

sbwsg commented on pkg/apis/pipeline/v1alpha1/param_types.go:
```diff
@@ -30,11 +42,97 @@ type ParamSpec struct {
+	if arrayOrString.Type == ParamTypeString {
+		return arrayOrString.StringVal
+	}
+	return strings.Join(arrayOrString.ArrayVal, " ")
```
Whereabouts is this used now?  I think you mentioned there was one spot in the code that uses this?

EliZucker replied:
Again, retyping my reply: 

You're right, that method isn't used! I removed it yesterday.

EliZucker commented:
Tagging @dlorenc because I heard you set up the code coverage reports: I'm not sure if it's something I'm doing wrong from my end, but the coverage reports clearly contradict what IntelliJ is reporting (and I think IntelliJ is right?). Specifically templating.go is reported for having much lower coverage than intellij reports (see my comparison [here](https://i.imgur.com/28t5Rpu.png)).

EliZucker commented:
Unholding this PR because, from my perspective, it's completely ready to merge (addressed feedback so far and I think the readme is clear).

/unhold

sbwsg commented:
/lgtm

EliZucker commented:
/hold cancel

EliZucker commented on pkg/apis/pipeline/v1alpha1/param_types_test.go:
```diff
@@ -0,0 +1,188 @@
+	}
+}
+
+type IntOrStringHolder struct {
```
fixed!

EliZucker commented on docs/tasks.md:
```diff
@@ -183,15 +186,17 @@ metadata:
     - name: build
       image: my-builder
-      args: ["build", "--flags=${inputs.params.flags}"]
+      args: ["build", "${inputs.params.flags}", "url=${someURL}"]
```
fixed!

EliZucker commented on docs/tasks.md:
````diff
@@ -183,15 +186,17 @@ metadata:
 ```
 
-The following `TaskRun` supplies a value for `flags`:
+The following `TaskRun` supplies a dynamic number of strings within the `flags` parameter:
````
fixed!

EliZucker commented on test/builder/param.go:
```diff
@@ -0,0 +1,32 @@
+
+// GenerateArrayOrString creates an ArrayOrString of type ParamTypeString or ParamTypeArray, based on
+// how many inputs are given (>1 input will create an array, not string).
+func GenerateArrayOrString(value string, additionalValues ...string) *v1alpha1.ArrayOrString {
```
Sure, makes sense. I'll change that right now then resolve this.

EliZucker commented:
/retest

EliZucker commented:
/test pull-tekton-pipeline-build-tests

EliZucker commented:
/retest

EliZucker commented:
/retest

afrittoli commented on pkg/reconciler/v1alpha1/pipelinerun/pipelinerun.go:
```diff
@@ -227,6 +231,21 @@ func (c *Reconciler) reconcile(ctx context.Context, pr *v1alpha1.PipelineRun) er
 		return nil
 	}
 
+	// Ensure that the parameters from the PipelineRun are overriding Pipeline parameters with the same type.
```
Just a though: I'm not quite sure how the separation of concerns might be between the validation webhook and the reconciler is meant to be exactly, but it feels to me that this is something that we could verify when the PipelineRun is sent to the API instead of accepting it a failing it here on the parameter types?
I don't mind if this stays like this, but if it makes sense at all we might consider moving this to the webhook.

EliZucker replied:
Oh interesting! In my head, I see webhook validation as validation that only uses information on that isolated YAML file being submitted (e.g invalid types, invalid variable interpolation strings), while reconciler validation brings in outside resources and evaluates the collective context. 

In this case, since the validation requires finding the referenced pipeline resource, I put it in the reconciler -- this particular test verifies that, for example, a pipelinerun parameter *string* value is not overriding a corresponding pipeline's parameter *array* value. This seems really similar to the validation done with `GetResourcesFromBindings` (`will validate that all PipelineResources declared in Pipeline p are bound in PipelineRun pr`), which is also in the reconciler.

Writing this up, I think my decision makes sense, but let me know if it doesn't!

EliZucker replied:
@afrittoli this is meant to be a reply to your comment above. GitHub is being weird though and showing it as a separate comment so I'm tagging you.

EliZucker commented on pkg/apis/pipeline/v1alpha1/param_types.go:
```diff
@@ -57,15 +57,15 @@ func (pp *ParamSpec) SetDefaults(ctx context.Context) {
 }
 
-// Param declares a value to use for the Param called Name, and is used in the
+// ResourceParam declares a value to use for the parameter called Name, and is used in the
```
fixed!

ImJasonH commented:
/lgtm
/approve

dibyom commented:
/retest

EliZucker commented:
@ImJasonH @afrittoli If one of you could give an lgtm that would be great! I had to rebase/refactor a bit to merge correctly with the initial conditional stuff, so the label was removed.

sbwsg commented:
:tada: Niiiiiiiiiice
